src/hotspot/share/memory/metaspaceShared.cpp

```diff
@@ -1612,34 +1613,50 @@ char* MetaspaceShared::reserve_address_space_for_archives(FileMapInfo* static_ma
+      address ccs_base = base_address + archive_space_size + gap_size;
+      archive_space_rs = ReservedSpace(archive_space_size, archive_space_alignment,
+                                       false /* large */, (char*)base_address);
+      class_space_rs   = ReservedSpace(class_space_size, class_space_alignment,
+                                       false /* large */, (char*)ccs_base);
+    }
+    if (!archive_space_rs.is_reserved() || !class_space_rs.is_reserved()) {
+      release_reserved_spaces(archive_space_rs, class_space_rs);
+      return NULL;
+    }
   } else {
-    // Reserve at any address, but leave it up to the platform to choose a good one.
-    total_rs = Metaspace::reserve_address_space_for_compressed_classes(total_range_size);
-  }
-
-  if (!total_rs.is_reserved()) {
-    return NULL;
-  }
-
-  // Paranoid checks:
-  assert(base_address == NULL || (address)total_rs.base() == base_address,
-         "Sanity (" PTR_FORMAT " vs " PTR_FORMAT ")", p2i(base_address), p2i(total_rs.base()));
-  assert(is_aligned(total_rs.base(), archive_space_alignment), "Sanity");
-  assert(total_rs.size() == total_range_size, "Sanity");
-  assert(CompressedKlassPointers::is_valid_base((address)total_rs.base()), "Sanity");
+    ReservedSpace total_rs;
+    if (base_address != NULL) {
+      // Reserve at the given archive base address, or not at all.
+      total_rs = ReservedSpace(total_range_size, archive_space_alignment,
+                               false /* bool large */, (char*) base_address);
+    } else {
+      // Reserve at any address, but leave it up to the platform to choose a good one.
+      total_rs = Metaspace::reserve_address_space_for_compressed_classes(total_range_size);
+    }
 
-  // Now split up the space into ccs and cds archive. For simplicity, just leave
-  //  the gap reserved at the end of the archive space.
-  archive_space_rs = total_rs.first_part(ccs_begin_offset,
-                                         (size_t)os::vm_allocation_granularity(),
-                                         /*split=*/true);
-  class_space_rs = total_rs.last_part(ccs_begin_offset);
+    if (!total_rs.is_reserved()) {
+      return NULL;
+    }
 
+    // Paranoid checks:
+    assert(base_address == NULL || (address)total_rs.base() == base_address,
+           "Sanity (" PTR_FORMAT " vs " PTR_FORMAT ")", p2i(base_address), p2i(total_rs.base()));
+    assert(is_aligned(total_rs.base(), archive_space_alignment), "Sanity");
+    assert(total_rs.size() == total_range_size, "Sanity");
+    assert(CompressedKlassPointers::is_valid_base((address)total_rs.base()), "Sanity");
+
+    // Now split up the space into ccs and cds archive. For simplicity, just leave
+    //  the gap reserved at the end of the archive space.
+    archive_space_rs = total_rs.first_part(ccs_begin_offset,
+                                           (size_t)os::vm_allocation_granularity(),
+                                           /*split=*/false);
```

Comment:
Since `/*split=*/false` is passed,  the region is no longer split using `os::split_reserved_memory()`. This means on Windows, we  cannot use `os::release_memory()` to free the individual regions of `archive_space_rs` and `class_space_rs`.

The Windows API of [VirtualFree](https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualfree) says:

> Releases the specified region of pages, or placeholder [...], `dwSize` must be 0 (zero), and `lpAddress` must point to the base address returned by the `VirtualAlloc` function when the region is reserved. The function fails if either of these conditions is not met.

I suggest we do this:

- Add an extra `ReservedSpace& total_rs` parameter to `MetaspaceShared::reserve_address_space_for_archives()`. Return the `total_rs` when we go through this path of the code.
- Also pass `total_rs` to `MetaspaceShared::release_reserved_spaces`. If `total_rs.is_reserverd()` is true, release `total_rs` instead of the two smaller spaces.

To make sure this PR is correct, we should add something like the following in `os::release_memory()`, and check for this log in test/hotspot/jtreg/runtime/cds/SharedBaseAddress.java:

```
if (!res) {
  log_info(os)("os::release_memory(" PTR_FORMAT ", " SIZE_FORMAT ") failed", p2i(addr), bytes);
}
```

Perhaps, in a separate RFE, we should add an assert in `os::release_memory()`, or at least change the `log_info` to `log_warning`.

Reply:
Hi Ioi,

Where is the release you worry about? Since on Windows, since 8255978, os::release_memory() notices if the region the caller wants release does not correspond exactly to a memory mapping at the OS level, and will assert. Do you see that assert?

I try to understand:

if useBaseAddress==true, Yumin now creates two separate mappings, and can release them individually
if useBaseAddress==false, there is one mapping as before, but we split now shallow. But we don't release it since we use file IO to read into it. If someone were to release one of those, we should see an assert on Windows.
(I am a tiny bit unhappy about the increasing complexity of the patch, since it negates some of the work done to simplify it back in June.)

About tracing, since 8256864 we trace Virtualxxx calls, so the tracing is already there - we trace VirtualFree() errrors for "os=info". If you only care for windows, that tracing could suffice.

Thanks, Thomas

Reply:
If `useBaseAddress==false`, after the total_space is successfully mapped, subsequent operations mail fail. For example,

- `mapinfo->map_regions()` may fail to commit the necessary memory for doing `os::read()`.
- `mapinfo->validate_shared_path_table()` may fail because the runtime classpath is not compatible

In these cases, we need to call MetaspaceShared::release_reserved_spaces().

> Do you see that assert?

Hmm, I think we should add a new test for this specifically: `java -XX:ArchiveRelocationMode=1 -cp mispatched.jar` to force the failure in `mapinfo->validate_shared_path_table()`.

>  If you only care for windows, that tracing could suffice.

Since we are changing the `split` parameter for all platforms, I think we should test for all platforms, not just windows.

Reply:
The new test case can be something like this:

```
$ java -XX:DumpLoadedClassList=HelloWorld.classlist -cp HelloWorld.jar HelloWorld
Hello World
$ java -Xshare:dump -XX:SharedClassListFile=HelloWorld.classlist -cp HelloWorld.jar \
       -XX:SharedArchiveFile=tmp.jsa -XX:SharedBaseAddress=0 
$ java -Xshare:auto -XX:SharedArchiveFile=tmp.jsa -cp XXX.jar:HelloWorld.jar -Xlog:cds \
       -showversion HelloWorld
...
[0.034s][info][cds] Archive(s) were created with -XX:SharedBaseAddress=0. Always map at os-selected address.
[0.034s][info][cds] Try to map archive(s) at an alternative address
[0.034s][info][cds] Mapped static  region #0 at base 0x00007f21f7800000 top 0x00007f21f7806000 (MiscCode)
[0.034s][info][cds] Mapped static  region #1 at base 0x00007f21f7806000 top 0x00007f21f7a1d000 (ReadWrite)
[0.034s][info][cds] Mapped static  region #2 at base 0x00007f21f7a1d000 top 0x00007f21f7daf000 (ReadOnly)
[0.048s][info][cds] UseSharedSpaces: shared class paths mismatch (hint: enable -Xlog:class+path=info to diagnose the failure)
[0.048s][info][cds] Unmapping region #0 at base 0x00007f21f7800000 (MiscCode)
[0.048s][info][cds] Unmapping region #1 at base 0x00007f21f7806000 (ReadWrite)
[0.048s][info][cds] Unmapping region #2 at base 0x00007f21f7a1d000 (ReadOnly)
[0.048s][info][cds] UseSharedSpaces: Unable to map shared spaces
java version "16-internal" 2021-03-16
Java(TM) SE Runtime Environment (slowdebug build 16-internal+0-adhoc.iklam.open)
Java HotSpot(TM) 64-Bit Server VM (slowdebug build 16-internal+0-adhoc.iklam.open, mixed mode)
Hello World
```
You can set a breakpoint at `MetaspaceShared::release_reserved_spaces()` to make sure it's called with `total_rs`.

Note that `-Xshare:auto` must be used. 

If `-Xshare:on` is used, the VM will exit immediately without calling `MetaspaceShared::release_reserved_spaces()`. Most of the CDS tests are executed with `-Xshare:on`. That's why we didn't see the assert on Windows with Yumin's earlier patch -- in Mach5 tier 4, we run test/hotspot/jtreg/runtime/cds/appcds/BootClassPathMismatch.java with `-XX:ArchiveRelocationMode=1`, but the test itself would use with `-Xshare:on`.

Reply:
When jar mismatched, say, -cp non-exist.jar (We could not use -cp no-exitst.jar:hello.jar since, once hello.jar the dump time jar found and matched, the non-exist.jar will be ignored, test showed it is OK), the output:
[0.035s][info ][cds] Mapped static  region #0 at base 0x00007f98e3800000 top 0x00007f98e3806000 (MiscCode)
[0.035s][info ][cds] Mapped static  region #1 at base 0x00007f98e3806000 top 0x00007f98e3a1e000 (ReadWrite)
[0.035s][info ][cds] Mapped static  region #2 at base 0x00007f98e3a1e000 top 0x00007f98e3db2000 (ReadOnly)
[0.049s][info ][cds] UseSharedSpaces: shared class paths mismatch (hint: enable -Xlog:class+path=info to diagnose the failure)
[0.049s][info ][cds] Unmapping region #0 at base 0x00007f98e3800000 (MiscCode)
[0.049s][info ][cds] Unmapping region #1 at base 0x00007f98e3806000 (ReadWrite)
[0.049s][info ][cds] Unmapping region #2 at base 0x00007f98e3a1e000 (ReadOnly)
[0.049s][debug][cds] Released shared space 0x00007f98e3800000
[0.049s][debug][cds] Released shared space (archive + class) 0x00007f98e3800000
[0.049s][info ][cds] UseSharedSpaces: Unable to map shared spaces

The repeat will be eliminated next version from here:
1489   } else {
1490     unmap_archive(static_mapinfo);
1491     unmap_archive(dynamic_mapinfo);
1492     log_debug(cds)("Released shared space " INTPTR_FORMAT, p2i(total_space_rs.base()));
1493     release_reserved_spaces(total_space_rs, archive_space_rs, class_space_rs);
1494   }

Reply:
You're correct that if you use `-cp non-exist.jar:hello.jar`, the non-exist.jar will be ignored because it doesn't exist, and CDS will not treat the classpath as a mismatch.

However, in my example, I created the XXX.jar file, so CDS will treat the classpath as a mismatch.